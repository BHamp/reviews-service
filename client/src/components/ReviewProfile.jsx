import React, { useState, useEffect } from 'react';

const ReviewProfile = (props) => {
  const { user } = props;
  return (
    <div className="leftcol">
      <div class="avatar">
        <img src={user.icon}></img>
      </div>
      <div className="username">
        {user.username}
      </div>
      <div className="num_owned_games">
        {`${user.numProducts} products in account`}
      </div>
      <div className="num_reviews">
        {`${user.numReviews} reviews`}
      </div>
    </div>
  )
};

export default ReviewProfile;