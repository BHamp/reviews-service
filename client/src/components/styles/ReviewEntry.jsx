import styled from 'styled-components';

export const Review = styled.div`
  border-bottom: 26px solid #1B2838;
`;

export const ReviewContent = styled.div`
  padding: 8px 8px 8px 16px;
  float: left;
  width: 400px;
  position: relative;
  width: 308px;
  margin-left: 14px;
  float: right;
`;

export const TooltipText = styled.span`
  visibility: hidden;
  right: 10px;
  position: absolute;
	padding: 0px 10px;
	box-shadow: 0 0 5px #000;
	background: -webkit-linear-gradient( top, #e3eaef 5%, #c7d5e0 95%);
	background: linear-gradient( to bottom, #e3eaef 5%, #c7d5e0 95%);
  z-index: 1;
  color:black;
  border-radius: 6px;
  text-align: center;
`;

export const VoteHeader = styled.div`
  margin: 8px 0 13px;
  display: block;
  background: rgba( 0, 0, 0, 0.2 );
  height: 40px;

  & ${TooltipText}:hover {
    visibility: visible;
    left: 50%;
    cursor: pointer;
  }

  &:hover {
    background: rgba( 255, 255, 255, 0.1 );
  }
`;

export const Thumb = styled.div`
  float: left;
  margin-right: 10px;
`;

export const ReviewSource = styled.div`
  float: right;
  margin-right: 5px;
  margin-top: 12px;
  opacity: 0.5;
`;

export const VoteHeaderTitle = styled.div`
  font-size: 16px;
  color: #d6d7d8;
  font-weight: normal;
  padding: 3px 0px 0px 0px;
  line-height: 19px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;

export const HoursPlayed = styled.div`
  font-weight: 300;
  font-size: 11px;
  line-height: 15px;
  color: #8091a2;
  opacity: 0.6;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;

export const DatePosted = styled.div`
  margin: 0px 0px 8px;
  font-size: 10px;
  text-transform: uppercase;
  font-style: normal;
  color: #8091a2;
  display: inline-block;
  opacity: 0.6;
`;

export const Content = styled.div`
  margin-right: 8px;
  font-weight: normal;
  font-size: 13px;
  line-height: 17px;
  color: #acb2b8;
  overflow-wrap: break-word;
  overflow: hidden;

  ${({ partial }) => partial && `
    max-height: 200px;
    overflow: visible;
    position: relative;
  `}

  ${({ expanded }) => expanded && `
    display: none;
  `}
`;

export const Gradient = styled.div`
  ${({ partial }) => partial && `
    display: block;
    position: absolute;
    bottom: 0px;
    left: 0px;
    right: 0px;
    z-index: 1;
    height: 30px;
    background: linear-gradient( to bottom, rgba( 22,32,45,0) 5%, rgba( 22,32,45,.95) 95%);
  `}

  ${({ expanded }) => expanded && `
    display: none;
  `}
`;

export const ViewMore = styled.div`
  ${({ partial }) => partial && `
    color: #67c1f5;
    text-transform: uppercase;
    font-style: normal;
    text-decoration: none;
    font-size: 10px;
    cursor: pointer;
  `}

  ${({ expanded }) => expanded && `
    display: none;
  `}
`;

export const Posted = styled.div`
  ${({ expanded }) => expanded && `
    font-size: 10px;
    font-style: italic;
    color: #626366;
    margin: 0px;
  `}
`;

export const Line = styled.div`
  background-color: #363f4c;
  height: 1px;
`;

export const ControlBlock = styled.div`
  margin: 8px 0;
  display: inline-block;
`;

export const Button = styled.button`
  border-radius: 2px;
  border: none;
  padding: 1px;
  display: inline-block;
  cursor: pointer;
  text-decoration: none;
  color: #66c0f4;
  margin: 16px;
  width: 16px;
  height: 16px;
  background: url("http://54.67.101.150:4200/img/votecontrolicons.png") no-repeat top left;
  vertical-align: text-top;
  display: table-cell;
  position: relative;

  &:hover {
    filter: saturate(4);
  }

  &:focus {
    background-color: #97d1f3;
    outline: none;
    box-shadow: none;
  }
`;

export const VoteUp = styled.span`
  padding: 0 5px;
  font-size: 12px;
  line-height: 20px;
  padding: 16px;
  background-position: -112px -16px;

  &:hover {
    background-position: -144px -16px;
  }
`;

export const VoteDown = styled.span`
  padding: 0 5px;
  font-size: 12px;
  line-height: 20px;
  padding: 16px;
  background-position: -64px -16px;

  &:hover {
    background-position: -80px -16px;
  }
`;

export const VoteFunny = styled.span`
  padding: 0 5px;
  font-size: 12px;
  line-height: 20px;
  padding: 16px;
  background-position: -208px -16px;

  &:hover {
    background-position: -224px -16px;
  }
`;


export const VoteInfo = styled.div`
  padding: 0px 9px 8px 0px;
  font-size: 12px;
  color: #647580;
  background-color: rgba( 0, 0, 0, 0.2 );
  min-height: 16px;

  & ${TooltipText}:hover {
    visibility: visible;
    left: 50%;
    cursor: pointer;
  }
`;

export const Break = styled.div`
  clear: left;
`;
