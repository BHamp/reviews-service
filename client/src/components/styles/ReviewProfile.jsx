import styled from 'styled-components';

export const Profile = styled.div`
  float: left;
  padding: 8px;
  opacity: 0.6;
`;

export const Avatar = styled.div`
  width: 34px;
  height: 34px;
  position: relative;
  border-radius: 0;
  padding: 1px;
  float: left;
  padding-right: 8px;
  display: block;
`;

export const AvatarImg = styled.img`
  width: 32px;
  height: 32px;
  padding: 1px;
  border-radius: 0;
  border: none;
`;

export const Username = styled.div`
  text-overflow: ellipsis;
  white-space: nowrap;
  width: 140px;
  overflow: hidden;
  margin-bottom: -1px;
  display: inline-block;
  color: #c1dbf4;
  font-size: 13px;
  font-weight: bold;
`;

export const NumOwnedGames = styled.div`
  font-size: 11px;
  color: #626366;
  display: block;
  line-height: 17px;

  &:hover {
    filter: invert(1);
  }
`;

export const NumReviews = styled.div`
  font-size: 11px;
  color: #626366;
  display: block;
  color: #c1dbf4;

  &:hover {
    filter: saturate(4);
  }
`;
