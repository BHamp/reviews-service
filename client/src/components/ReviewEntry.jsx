import React from 'react';
import ReviewProfile from './ReviewProfile.jsx';

const ReviewEntry = (props) => {
  const { review, updateVote } = props;

  let date = new Date(review.date_posted);
  let options = { month: 'long', day: 'numeric' };
  let postedDate = date.toLocaleString('en-US', options);

  const handleContent = (e) => {
    const element = document.getElementById('ReviewContent');
    element.classList.remove('partial');
    element.classList.add('expanded');
  };

  const handleVote = (e) => {
    let value = review.helpful;
    if (e.target.id === 'voteFunny') {
      value = review.funny;
    }
    updateVote(e.target.id, value, review.user.id);
  };

  return (
    <div id="ReviewContent"  className="partial">
      <ReviewProfile user={review.user} key={review.user.id} />
      <div className="rightcol">
        <div className="vote_header">
          <div className="thumb">
            <img src="./img/thumbsup.png" width="40" height="40" ></img>
          </div>
          <img className="review_source" src="./img/steamicon.png"></img>
          <div className="title ellipsis">Recommended</div>
          <div className="hours ellipsis">
            {`${review.hours} gametime hours on record`}
          </div>
        </div>
        <div className="postedDate">
          {`Posted: ${postedDate}`}
        </div>
        <div className="content">
          {review.description}
          <div className="gradient" />
        </div>
        <div className="posted">
          <div className="view_more">
            <a onClick={handleContent} >Read More</a>
          </div>
          <div className="line" />
        </div>
        <div className="control_block">
          Was this review helpful?
          <br />
          <button
            type="button"
            className="votecontrol voteUp"
            onClick={handleVote}
          >
            <span id="voteUp">Yes</span>
          </button>
          <button
            type="button"
            className="votecontrol voteDown"
            onClick={handleVote}
          >
            <span id="voteDown">No</span>
          </button>
          <button
            type="button"
            className="votecontrol voteFunny"
            onClick={handleVote}
          >
            <span id="voteFunny">Funny</span>
          </button>
        </div>
        <div className="vote_info">
          {`${review.helpful} people found this review helpful`}
          <br />
          {`${review.funny} people found this review funny`}
          <br />
          {`${review.thread_length} additional comments`}
          <br />
        </div>
      </div>
      <div style={{ clear: 'left' }} />
    </div>
  );
};

export default ReviewEntry;
