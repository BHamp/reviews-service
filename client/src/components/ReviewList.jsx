import React, { useState, useEffect } from 'react';
import ReviewEntry from './ReviewEntry.jsx';

const ReviewList = (props) => {
  const { allReviews } = props;

  return (
    <div className="review_box">
      {allReviews.map((review) => {
        return <ReviewEntry review={review} />;
      })}
    </div>
  );
};


export default ReviewList;
