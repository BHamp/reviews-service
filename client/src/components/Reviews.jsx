import React, { useState, useEffect} from 'react';
import axios from 'axios';
import ReviewList from './ReviewList.jsx';
import ReviewSummaryBar from './ReviewSummaryBar.jsx';

const Reviews = () => {
  const [allReviews, setAllReviews] = useState([]);
  let id = window.location.search.substring(2);

  const getAllReviews = () => {
    axios.get(`http://localhost:4200/api/reviews/${id}`)
      .then(({ data }) => {
        setAllReviews(data);
      })
      .catch((err) => {
        console.error(err);
      });
  };

  useEffect(() => {
    getAllReviews();
  }, []);

  const updateVotes = (e, val, reviewId) => {
    let field = 'helpful';
    if (e === 'voteUp') {
      val += 1;
    } else if (e === 'voteDown') {
      val -= 1;
    } else {
      val += 1;
      field = 'funny';
    }

    axios.patch(`http://localhost:4200/api/reviews/${id}`, {
      field: field,
      value: val,
      reviewId: reviewId,
    })
      .then((results) => {
        getAllReviews();
      })
      .catch((err) => {
        console.error(err);
      });
  };

  return (
    <div className="user_reviews">
      <h2>Customer Reviews</h2>
      <ReviewSummaryBar allReviews={allReviews} />
      <div className="left-col">
        <div className="user_reviews_sub_header">
          {`Most Helpful Reviews `}
          <span className="user_reviews_most_helpful_days">
            {`In the past 30 days`}
          </span>
        </div>
        <ReviewList allReviews={allReviews} key={id} handleVote={updateVotes} />
      </div>
    </div>
  );
};

export default Reviews;
