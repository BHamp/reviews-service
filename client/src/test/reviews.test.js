import React from 'react';
import Enzyme, {
  configure, shallow, mount,
} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Reviews from '../components/Reviews';
import ReviewList from '../components/ReviewList';
import ReviewEntry from '../components/ReviewEntry';
import ReviewProfile from '../components/ReviewProfile';

import exampleData from './reviews.example';


configure({ adapter: new Adapter() });

describe('Renders components successfully', () => {
  test('should render the Reviews component', () => {
    expect(shallow(<Reviews />).find('.left-col').exists()).toBe(true);
  });

  test('loads ReviewList component', () => {
    mount(<Reviews />);
    expect(shallow(<ReviewList allReviews={exampleData} />).exists()).toBe(true);
  });

  test('loads ReviewEntry component', () => {
    mount(<ReviewList allReviews={exampleData} />);
    expect(shallow(<ReviewEntry review={exampleData[0]} />).exists()).toBe(true);
  });

});
