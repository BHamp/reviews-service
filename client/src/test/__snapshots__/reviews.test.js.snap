// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reviews should render correctly in "debug" mode 1`] = `
<div
  className="left-col"
>
  <div
    className="user_reviews_sub_header"
  >
    Most Helpful Reviews 
    <span
      className="user_reviews_most_helpful_days"
    >
      In the past 30 days
    </span>
  </div>
  <ReviewList
    allReviews={Array []}
    handleVote={[Function]}
    key=""
  />
</div>
`;
